### PR TITLE
infer right side nullability for LEFT join

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes partially #5747 (LEFT Join Mismatch between schema and batches on a CREATE TABLE with a windowing query  #5695 (comment)).

# Rationale for this change
Weakens the schema comparison between planner schema and batches leaving only name and datatype to avoid false positives on checking not null and nullable columns

# What changes are included in this PR?

# Are these changes tested?
yes

# Are there any user-facing changes?

### CODE REVIEW

**Comment by milevin** [comment]
@comphead I am not sure about this change; it doesn't seem sound. 

The analyzer is assigning incorrect nullability flags to the right side of the join. I wonder if we shouldn't fix this instead.

**Comment by comphead** [comment]
I think its not that easy for analyzer to identify correct nullable flags without scanning entire data. for LEFT join it assumes right side cannot be matched and sets nullable and this sounds correct for me. 

From other side we also have similar weaken check in optimizer https://github.com/apache/arrow-datafusion/blob/26b8377b0690916deacf401097d688699026b8fb/datafusion/optimizer/src/optimizer.rs#L423

**Comment by milevin** [comment]
Consider the above query. After `plan_selection` is finished in `select.rs`, the assigned column properties are as follows:

```
        DFField {
            qualifier: Some(
                Bare {
                    table: "t1",
                },
            ),
            field: Field {
                name: "col1",
                data_type: Int64,
                nullable: false,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        },
        DFField {
            qualifier: Some(
                Bare {
                    table: "t1",
                },
            ),
            field: Field {
                name: "col2",
                data_type: Utf8,
                nullable: false,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        },
        DFField {
            qualifier: Some(
                Bare {
                    table: "t2",
                },
            ),
            field: Field {
                name: "col3",
                data_type: Int64,
                nullable: false,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        },
        DFField {
            qualifier: Some(
                Bare {
                    table: "t2",
                },
            ),
            field: Field {
                name: "col4",
                data_type: Utf8,
                nullable: false,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        },
```

The two `t2` fields have incorrect nullability flag inferred. Unless I am missing something, the `plan_selection` method does not need to scan the data to infer the correct nullability flags. All it needs is the understanding of which join it is processing and which side of that join the columns are coming from.

Does it make sense?

**Comment by comphead** [comment]
yes, good catch, col4 has to have a true nullability

**Comment by milevin** [comment]
This is before optimizer I believe. It's just the initial stage of building the logical plan from the query.

**Comment by milevin** [comment]
This is great! Check out the comments above.

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -1041,20 +1041,44 @@ pub fn build_join_schema(
...
+        JoinType::Inner | JoinType::Full | JoinType::Right => {
             // left then right
-            left_fields.chain(right_fields).cloned().collect()
+            left_fields
+                .iter()
+                .chain(right_fields.iter())
+                .cloned()
+                .collect()
+        }
+        JoinType::Left => {
```
Comment (comphead) [inline datafusion/expr/src/logical_plan/builder.rs:1056]: I'll fix other types of joins on separate PR

File: datafusion/common/src/utils.rs [hunk 2]
```diff
@@ -234,6 +235,18 @@ pub(crate) fn parse_identifiers_normalized(s: &str) -> Vec<String> {
         .collect::<Vec<_>>()
 }
 
+/// Check two schemas for being equal for field names/types
+pub fn equivalent_names_and_types(schema: &SchemaRef, other: SchemaRef) -> bool {
```
[the remark this answers is not in the document]
Comment (milevin) [reply datafusion/common/src/utils.rs:239]: @comphead is this change relevant here?

**Review by milevin — APPROVED** [review]

File: datafusion/common/src/utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/common/src/utils.rs]: Thanks, done, Ive got a better idea how to refactor this method.

**Comment by comphead** [comment]
@alamb @jackwener please check this PR so I can follow up later on other join types

File: datafusion/expr/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -1041,20 +1041,44 @@ pub fn build_join_schema(
...
+            // left then right, right set to nullable in case of not matched scenario
+            let right_fields_nullable: Vec<DFField> = right_fields
+                .iter()
+                .map(|f| {
+                    let field = f.field().clone().with_nullable(true);
+                    if let Some(q) = f.qualifier() {
+                        DFField::from_qualified(q, field)
+                    } else {
+                        DFField::from(field)
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1065]: Maybe we could add a `DFField::with_nullable` to make this more concise:

```
                    let field = f.clone().with_nullable(true);
```

**Review by alamb — APPROVED** [review]
This change makes sense to me -- thank you @comphead 

It is somewhat amazing that his wasn't caught by existing tests.

**Comment by alamb** [comment]
> The analyzer is assigning incorrect nullability flags to the right side of the join. I wonder if we shouldn't fix this inst

I believe the story is that @comphead plans to fix this in a follow on PR. Is that accurate @comphead ?

Thank you for the review @milevin

**Comment by milevin** [comment]
> > The analyzer is assigning incorrect nullability flags to the right side of the join. I wonder if we shouldn't fix this inst
> 
> I believe the story is that @comphead plans to fix this in a follow on PR. Is that accurate @comphead ?

@alamb This comment of mine was on an original version of this PR which tried to address the issue by weakening the compatibility check; @comphead followed the suggestion in the comment to rewrite the fix by changing the analyzer instead -- so all is good! 

Note: for the case of LEFT join, the right side must be nullified (that's what the comment talks about). A follow on PR will handle RIGHT and OUTER join where the left side will have to be adjusted as well.

File: datafusion/expr/src/logical_plan/builder.rs (hunk 3 above)
Comment (comphead) [reply datafusion/expr/src/logical_plan/builder.rs:1065]: That sounds great! I'll create another PR to extend API

**Comment by comphead** [comment]
> > > The analyzer is assigning incorrect nullability flags to the right side of the join. I wonder if we shouldn't fix this inst
> > 
> > 
> > I believe the story is that @comphead plans to fix this in a follow on PR. Is that accurate @comphead ?
> 
> @alamb This comment of mine was on an original version of this PR which tried to address the issue by weakening the compatibility check; @comphead followed the suggestion in the comment to rewrite the fix by changing the analyzer instead -- so all is good!
> 
> Note: for the case of LEFT join, the right side must be nullified (that's what the comment talks about). A follow on PR will handle RIGHT and FULL OUTER joins where the left side will have to be adjusted as well.
> 
> -- Michael

Right, for RIGHT/FULL OUTER joins I will create a follow up PR. This one is too overloaded :)

**Comment by alamb** [comment]
Thanks all !

**Pull request merged**